### PR TITLE
CR for 20606 - If an external texture isn't found, still display the model

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
Default textures are substituted at run-time until the texture file is found and loaded.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8ecefdfe392931be3cab7fdbebf1ae0239a3d816 available at: https://gist.github.com/ea87d114ae21171f2748<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d44743563b531ff99e5fbd8549a2504313ce469a available at: https://gist.github.com/ea87d114ae21171f2748<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.dmg

**Comment by Menithal** [comment]
Heya, I tested this build. Object now loads, even if the texture is missing: however there is no indication or texture in the textures place that would indicate that there was an issue loading the texture. In my Opinion,  it probably would good to a have some sort of error texture for it.

**Comment by ctrlaltdavid** [comment]
There not being an error texture is intentional: the model FBX is displayed as soon as it is downloaded: it is displayed as grey and then coloured in with textures (diffuse, normal, specular, and emissive - each independently) as and if they download. Thus the model "rezzes in" ASAP without waiting for laggardly or nonexistent textures. An error is logged in the program log for each texture that can't be downloaded.

File: libraries/render-utils/src/Model.cpp [hunk 1]
```diff
@@ -2049,16 +2047,15 @@ void Model::renderPart(RenderArgs* args, int meshIndex, int partIndex, bool tran
             }
 
             if (!mesh.tangents.isEmpty()) {                 
-                Texture* normalMap = networkPart.normalTexture.data();
-                batch.setResourceTexture(1, !normalMap ?
-                    textureCache->getBlueTexture() : normalMap->getGPUTexture());
-
+                NetworkTexture* normalMap = networkPart.normalTexture.data();
+                batch.setResourceTexture(1, !normalMap || !normalMap->isLoaded() ?
```
Comment (Atlante45) [inline libraries/render-utils/src/Model.cpp]: Can you put parenthesis around `!normalMap || !normalMap->isLoaded()` makes it more obvious what's going on.

File: libraries/render-utils/src/Model.cpp [hunk 2]
```diff
@@ -2049,16 +2047,15 @@ void Model::renderPart(RenderArgs* args, int meshIndex, int partIndex, bool tran
...
+                                        textureCache->getBlueTexture() : normalMap->getGPUTexture());
             }
     
             if (locations->specularTextureUnit >= 0) {
-                Texture* specularMap = networkPart.specularTexture.data();
-                batch.setResourceTexture(locations->specularTextureUnit, !specularMap ?
-                                            textureCache->getWhiteTexture() : specularMap->getGPUTexture());
+                NetworkTexture* specularMap = networkPart.specularTexture.data();
+                batch.setResourceTexture(locations->specularTextureUnit, !specularMap || !specularMap->isLoaded() ?
```
Comment (Atlante45) [inline libraries/render-utils/src/Model.cpp]: Same thing here

File: libraries/render-utils/src/Model.cpp [hunk 3]
```diff
@@ -2073,9 +2070,9 @@ void Model::renderPart(RenderArgs* args, int meshIndex, int partIndex, bool tran
                 float emissiveScale = part.emissiveParams.y;
                 GLBATCH(glUniform2f)(locations->emissiveParams, emissiveOffset, emissiveScale);
                 
-                Texture* emissiveMap = networkPart.emissiveTexture.data();
-                batch.setResourceTexture(locations->emissiveTextureUnit, !emissiveMap ?
-                                        textureCache->getWhiteTexture() : emissiveMap->getGPUTexture());
+                NetworkTexture* emissiveMap = networkPart.emissiveTexture.data();
+                batch.setResourceTexture(locations->emissiveTextureUnit, !emissiveMap || !emissiveMap->isLoaded() ?
```
Comment (Atlante45) [inline libraries/render-utils/src/Model.cpp]: Same thing here.

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -2049,16 +2047,15 @@ void Model::renderPart(RenderArgs* args, int meshIndex, int partIndex, bool tran
...
+                                        textureCache->getBlueTexture() : normalMap->getGPUTexture());
             }
     
             if (locations->specularTextureUnit >= 0) {
-                Texture* specularMap = networkPart.specularTexture.data();
-                batch.setResourceTexture(locations->specularTextureUnit, !specularMap ?
-                                            textureCache->getWhiteTexture() : specularMap->getGPUTexture());
+                NetworkTexture* specularMap = networkPart.specularTexture.data();
+                batch.setResourceTexture(locations->specularTextureUnit, !specularMap || !specularMap->isLoaded() ?
+                                        textureCache->getBlackTexture() : specularMap->getGPUTexture());
```
Comment (Atlante45) [inline libraries/render-utils/src/Model.cpp]: Shouldn't the default specular be white? (cc @samcake)

Comment (ctrlaltdavid) [inline libraries/render-utils/src/Model.cpp]: White make an object with missing specular texture very shiny.

Comment (Atlante45) [inline libraries/render-utils/src/Model.cpp]: Yup, pretty much what Sam just told me.
He said he'd check the code though so I'll wait for his green light.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2bab7d1c037845c87dbc18b521d546bf361a175a available at: https://gist.github.com/ea87d114ae21171f2748<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5306.dmg

**Pull request merged**